pkg/controllers/nodeclass/status/nodeimage.go

```diff
@@ -0,0 +1,179 @@
+			RateLimiter:             reasonable.RateLimiter(),
+			MaxConcurrentReconciles: 10,
+		}).
+		Complete(reconcile.AsReconciler(m.GetClient(), r))
+}
+
+// The image version reconciler will detect reasons to bump the node image version as follows in order:
+//
+// Scenario A: Update all image versions to latest
+//   - 1. Initializes the images versions for a newly created AKSNodeClass, based on customer configuration.
+//   - 2. Indirectly handle image bump for k8s upgrade
+//   - 3. Can indirectly handle bumps for any images unsupported by node features, if required to in the future
+//     Note: Currently there are no node features to be handled in this way.
+//   - 4. TODO: Update NodeImages to latest if in a MW (maintenance windows) [retrieved from ConfigMap]
+//
+// Scenario B: Calculate images to be updated based on delta of available images
+//   - 5. Handles update cases when customer changes image family, SIG usage, or other means of image selectors
+//   - 6. Handles softly adding newest image version of any newly supported SKUs by Karpenter
+func (r *NodeImageReconciler) Reconcile(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass) (reconcile.Result, error) {
+	logger := logging.FromContext(ctx)
+	logger.Debug("nodeclass.nodeimage: starting reconcile")
+
+	nodeImages, err := r.nodeImageProvider.List(ctx, nodeClass)
+	if err != nil {
+		logger.Debug("nodeclass.nodeimage: err listing node images")
+		return reconcile.Result{}, fmt.Errorf("getting nodeimages, %w", err)
+	}
+	images := lo.Map(nodeImages, func(nodeImage imagefamily.NodeImage, _ int) v1alpha2.NodeImage {
+		reqs := lo.Map(nodeImage.Requirements.NodeSelectorRequirements(), func(item v1.NodeSelectorRequirementWithMinValues, _ int) corev1.NodeSelectorRequirement {
+			return item.NodeSelectorRequirement
+		})
+
+		sort.Slice(reqs, func(i, j int) bool {
+			if len(reqs[i].Key) != len(reqs[j].Key) {
+				return len(reqs[i].Key) < len(reqs[j].Key)
+			}
+			return reqs[i].Key < reqs[j].Key
+		})
+		return v1alpha2.NodeImage{
+			ID:           nodeImage.ID,
+			Requirements: reqs,
+		}
+	})
+
+	// Scenario A: Check if we should do a full update to latest before processing any partial update
+	shouldUpdate := imageVersionsUnready(nodeClass) || isOpenMW()
+	logger.Debugf("nodeclass.nodeimage: should complete a full update to latest: %t", shouldUpdate)
+	if !shouldUpdate {
+		// Scenario B: Check if we should do any partial update based on image selectors, or newly supports SKUs
+		images, shouldUpdate = processPartialUpdate(nodeClass, images)
+	}
+
+	logger.Debugf("nodeclass.nodeimage: should update overall: %t", shouldUpdate)
+	if shouldUpdate {
+		if len(images) == 0 {
+			nodeClass.Status.NodeImages = nil
+			nodeClass.StatusConditions().SetFalse(v1alpha2.ConditionTypeNodeImageReady, "NodeImagesNotFound", "NodeImageSelectors did not match any NodeImages")
+			logger.Info("nodeclass.nodeimage: no images")
+			return reconcile.Result{RequeueAfter: time.Minute}, nil
```

Comment:
minor: Why a different interval than below?

Reply:
Will update to match